src/core/namespace.c

```diff
@@ -3155,6 +3157,118 @@ bool ns_type_supported(NamespaceType type) {
+                FOREACH_ARRAY(m, ml.mounts, ml.n_mounts) {
+                        if (m->mode == MOUNT_EXTENSION_DIRECTORY || m->mode == MOUNT_EXTENSION_IMAGE) {
+                                r = apply_one_mount(p->root_directory, m, p);
+                                if (r < 0) {
+                                        log_debug_errno(r, "Failed to apply extension mount: %m");
+                                        _exit(EXIT_FAILURE);
+                                }
+                        } else if (m->mode == MOUNT_OVERLAY) {
+                                _cleanup_free_ char *path_relative = NULL, *path_in_namespace = NULL;
+
+                                r = mount_overlay(m);
+                                if (r < 0)
+                                        _exit(EXIT_FAILURE);
+                                if (r == 0) /* Tried to mount overlay, but skipped. */
+                                        continue;
+
+                                /* bind_mount_in_namespace takes a src on the outside
+                                * and a dest evaluated within the namespace. First,
+                                * figure out where we want the overlay on top of within
+                                * the namespace.
+                                */
+                                r = path_make_relative(overlay_prefix, mount_entry_path(m), &path_relative);
+                                if (r < 0) {
+                                        log_debug_errno(r, "Failed to make path relative: %m");
+                                        _exit(EXIT_FAILURE);
+                                }
+                                r = asprintf(&path_in_namespace, "%s/%s", empty_to_root(p->root_directory), path_relative);
+                                if (r < 0) {
+                                        log_oom_debug();
+                                        _exit(EXIT_FAILURE);
+                                }
+
+                                r = bind_mount_in_namespace(
+                                                target,
+                                                p->propagate_dir,
+                                                p->incoming_dir,
+                                                /* src= */ mount_entry_path(m),
+                                                /* dest= */ path_in_namespace,
+                                                min_flags);
+                                if (r < 0) {
+                                        log_debug_errno(
+                                                        r,
+                                                        "Failed to move overlay within %s->%s: %m",
+                                                        mount_entry_path(m),
+                                                        path_in_namespace);
+                                        _exit(EXIT_FAILURE);
+                                }
```

Comment:
this really should be using mount_exchange_graceful() so that we can atomcially replace the old overlayfs with the new one, and do not collect unbounded oerlayfs.

Reply:
I've kept it using `mount_in_namespace` for now, because it does try to do `mount_exchange_graceful` under the hood (and we are also using it to handle umounting, with bluca's patch below), but I can revisit this again.

Reply:
Yeah this already does the swap using mount beneath, so it should be fine